jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/UsageResource.java

```diff
@@ -136,7 +133,6 @@ public Response recordUsage(final SubscriptionUsageRecordJson json,
         final Entitlement entitlement = entitlementApi.getEntitlementForId(json.getSubscriptionId(), false, callContext);
         if (entitlement.getEffectiveEndDate() != null) {
             final DateTime highestRecordDate = getHighestRecordDate(json.getUnitUsageRecords());
-            final Account account = accountUserApi.getAccountById(entitlement.getAccountId(), callContext);
```

Comment:
I would rename `callContext` to `callContextNoAccountId` and recreate `callContext` with the `accountId` right after retrieving `entitlement` (we use this pattern in a few other places in jax-rs).

Reply:
@pierre Do you mind explaining why? This commit is about removing this line:
```java
final Account account = accountUserApi.getAccountById(entitlement.getAccountId(), callContext)
```
Because `account` not used anymore. FYI:

1. that `account` introduced by [this commit](https://github.com/killbill/killbill/commit/5ded768eaf13082d107eec92d3fdc7c7ed716392#diff-01bacc5f459c602e5c3c753b9f99dc8a6f47c5e8cf5458e97025f451e2264c25R138) (because `account` attributes needed to create `TimeAwareContext` instance).
2. And then, in sub-sequence commit (08f9b913) the `TimeAwareContext` [get removed](https://github.com/killbill/killbill/commit/08f9b913b886b8e497b6edde02d51f4166137e69#diff-01bacc5f459c602e5c3c753b9f99dc8a6f47c5e8cf5458e97025f451e2264c25L139), but `account` still there, and have no use.

Or, do you mean that `accountUserApi.getAccountById(entitlement.getAccountId(), callContext)` should be called no matter what because it manage/changing any state? :thinking:

Reply:
Actually, you don't need to retrieve the account. I am suggesting doing something like that: 

```
diff --git a/jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/UsageResource.java b/jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/UsageResource.java
index 1546e86d48..c2caa820f1 100644
--- a/jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/UsageResource.java
+++ b/jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/UsageResource.java
@@ -131,9 +131,12 @@ public class UsageResource extends JaxRsResourceBase {
                 verifyNonNull(usageRecordJson.getRecordDate(), "UsageRecordJson recordDate needs to be set");
             }
         }
-        final CallContext callContext = context.createCallContextNoAccountId(createdBy, reason, comment, request);
+        final CallContext callContextNoAccountId = context.createCallContextNoAccountId(createdBy, reason, comment, request);
+
         // Verify subscription exists..
-        final Entitlement entitlement = entitlementApi.getEntitlementForId(json.getSubscriptionId(), false, callContext);
+        final Entitlement entitlement = entitlementApi.getEntitlementForId(json.getSubscriptionId(), false, callContextNoAccountId);
+        final CallContext callContext = context.createCallContextWithAccountId(entitlement.getAccountId(), createdBy, reason, comment, request);
+
         if (entitlement.getEffectiveEndDate() != null) {
             final DateTime highestRecordDate = getHighestRecordDate(json.getUnitUsageRecords());
             final Account account = accountUserApi.getAccountById(entitlement.getAccountId(), callContext);
```

In general, the context should always have the accountId populated (this is used in other layers below). But sometimes we cannot construct it directly (like here) and we need to construct it in multiple stages. Another example:
https://github.com/killbill/killbill/blob/3a52a09da73c4e6d06cdb23272e5c070789ca345/jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/InvoicePaymentResource.java#L186-L189

Reply:
Ah, so in `UsageResource.recordUsage()` use case, the `callContext` that [passed to usageUserApi.recordRolledUpUsage()](https://github.com/killbill/killbill/blob/09781d15977a513975945a75e31998317a1ac2e1/jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/UsageResource.java#LL142C22-L142C41) need to contains valid `accountId`. This is the body of `UsageResource.recordUsage()` method (validation part removed):

```java
        final CallContext callContextNoAccount = context.createCallContextNoAccountId(createdBy, reason, comment, request);
        // Verify subscription exists..
        final Entitlement entitlement = entitlementApi.getEntitlementForId(json.getSubscriptionId(), false, callContextNoAccount);
        if (entitlement.getEffectiveEndDate() != null) {
            final DateTime highestRecordDate = getHighestRecordDate(json.getUnitUsageRecords());
            if (entitlement.getEffectiveEndDate().compareTo(highestRecordDate) < 0) {
                return Response.status(Status.BAD_REQUEST).build();
            }
        }

        final SubscriptionUsageRecord record = json.toSubscriptionUsageRecord();
        // See this discussion: https://github.com/killbill/killbill/pull/1861#discussion_r1199581600
        final CallContext callContext = context.createCallContextWithAccountId(entitlement.getAccountId(), createdBy, reason, comment, request);
        usageUserApi.recordRolledUpUsage(record, callContext);
        return Response.status(Status.CREATED).build();
```

Is this correct?

Reply:
Yes, that works (my suggested patch above was similar). 👍